git_machete/cmd.py

```diff
@@ -1094,20 +1102,81 @@ def merge_base(cli_ctxt: CommandLineContext, sha1: str, sha2: str) -> str:
 
 # Note: the 'git rev-parse --verify' validation is not performed in case for either of earlier/later
 # if the corresponding prefix is empty AND the revision is a 40 hex digit hash.
-def is_ancestor(cli_ctxt: CommandLineContext, earlier_revision: str, later_revision: str, earlier_prefix: str = "refs/heads/", later_prefix: str = "refs/heads/") -> bool:
-    if earlier_prefix == "" and is_full_sha(earlier_revision):
-        earlier_sha: Optional[str] = earlier_revision
-    else:
-        earlier_sha = commit_sha_by_revision(cli_ctxt, earlier_revision, earlier_prefix)
-    if later_prefix == "" and is_full_sha(later_revision):
-        later_sha: Optional[str] = later_revision
-    else:
-        later_sha = commit_sha_by_revision(cli_ctxt, later_revision, later_prefix)
+def is_ancestor(
+    cli_ctxt: CommandLineContext,
+    earlier_revision: str,
+    later_revision: str,
+    earlier_prefix: str = "refs/heads/",
+    later_prefix: str = "refs/heads/",
+) -> bool:
+    earlier_sha = full_sha(cli_ctxt, earlier_revision, earlier_prefix)
+    later_sha = full_sha(cli_ctxt, later_revision, later_prefix)
+
     if earlier_sha == later_sha:
         return True
     return merge_base(cli_ctxt, earlier_sha, later_sha) == earlier_sha
 
 
+# Determine if any revisions in later_revision contain a tree with identical
+# contents to earlier_revision, indicating that later_revision contains a
+# rebase or squash merge of earlier_revision.
+def contains_equivalent_tree(
+    cli_ctxt: CommandLineContext,
+    earlier_revision: str,
+    later_revision: str,
+    earlier_prefix: str = "refs/heads/",
+    later_prefix: str = "refs/heads/",
+) -> bool:
+    earlier_sha = full_sha(cli_ctxt, earlier_revision, earlier_prefix)
+    later_sha = full_sha(cli_ctxt, later_revision, later_prefix)
+
+    if earlier_sha == later_sha:
+        return True
+
+    debug(
+        cli_ctxt,
+        "contains_equivalent_tree",
+        f"earlier_revision={earlier_revision} later_revision={later_revision}",
+    )
+
+    # git rev-list later_sha ^earlier_sha
+    # shows all commits reachable via later_sha but not by earlier_sha
+    #
+    # https://git-scm.com/docs/git-rev-list#_description
+    intermediate_shas = non_empty_lines(
+        popen_git(
+            cli_ctxt,
+            "rev-list",
+            later_sha,
+            "^" + earlier_sha,
+        )
+    )
+
+    for intermediate_sha in intermediate_shas:
+        # git diff-tree --quiet
+        # Exits with 1 if there were differences and 0 means no differences.
+        #
+        # https://git-scm.com/docs/git-diff-tree#Documentation/git-diff-tree.txt---quiet
+        diff_result = run_cmd(
+            cli_ctxt,
+            "git",
+            "diff-tree",
```

Comment:
Wow there's quite a lot of these `diff-tree` calls 😯 

```
➜ 18:59 ~/git-machete detect_squash PR #129 (asford) $ git machete status -v
git for-each-ref --format=%\(refname\)$'\t'%\(objectname\)$'\t'%\(committerdate:raw\) refs/remotes
git for-each-ref --format=%\(refname\)$'\t'%\(objectname\)$'\t'%\(committerdate:raw\)$'\t'%\(upstream\) refs/heads
git rev-list 0ec951c2cee12195b2cbdfdf131a79abf8974620 ^f59c47a52e86749d522472a0920dcbb7f7bca88c
git merge-base 0ec951c2cee12195b2cbdfdf131a79abf8974620 f59c47a52e86749d522472a0920dcbb7f7bca88c
git config --list
git rev-parse --verify --quiet 6fd151979d1684c70fb1887e9ad19b70c758c137^{commit}
git rev-parse --verify --quiet 8a509d9bde56d9db60607c22e7f2726cf5532c82^{commit}
git merge-base 6fd151979d1684c70fb1887e9ad19b70c758c137 8a509d9bde56d9db60607c22e7f2726cf5532c82
git merge-base 8a509d9bde56d9db60607c22e7f2726cf5532c82 f59c47a52e86749d522472a0920dcbb7f7bca88c
git rev-list f59c47a52e86749d522472a0920dcbb7f7bca88c ^73021bc11a37c530a892cbea1bd25d6f986b0c21
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 f59c47a52e86749d522472a0920dcbb7f7bca88c
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 b9b971e1007d3f7c2ddfb309da36c5e8d855ec37
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 a9424c5c1c1a883e7ed776b7c38f6e62d31fdad5
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 0f217af0f99884acba395cee61a83f9c4b8234dd
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 2aae74329af314c95f001dc6c910f231f650563a
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 76543acef3dca7492c28bcd4382f68cdd8e562b2
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 bf802772ca30cc02c85dde555e580a9deb7f6a65
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 3b6c90b1d44333c7678e2052ee96c7cf4a82e23a
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 895073cacd9d7c474381dac5e4486ecec9cc7fda
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 032230e12fbdcc612e417bc69613f51f82a6bacb
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 acc315cb678da3a17a5b8d27a1986770b773037b
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 0ec951c2cee12195b2cbdfdf131a79abf8974620
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 0bfab55b786b835e3514960f45a06876b5718bd2
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 a6429a4cc94dae48029c22b155bac8ed741f2d97
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 51650f5b0e590e55526a831c5e49cb7e00ef3f67
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 5750fa663d355619df0533e8f5bd2d8219d58a7f
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 3cdeb01903b6f9ecaae834cb25f81b9371d5bcec
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 5947d6ebdc4fd95ecf8e8f9769e235b21528a27f
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 8a435e0bb3e8aff7f59dfb7e925bc6ad80d70254
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 465f6d36458dd9f901549a911d461021f363f80c
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 7de2a1fd127c4eb62a9a649871942c0410f195a9
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 fa29c47d5d04615c3e20103a8174cd0a3b508851
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 2b762c7895e6b5dc32f56d6696b8c13361c1ce62
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 7afc223e47e31c461d5521269c76eb259914d77a
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 b27619ddf90cbdb3ca4d4bafd5248f8a2c38ce49
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 162e5b84c9f1741c5883f98f9f550b9b6fe7efc5
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 be4baf722ca32bd317f954374c3da6d08dae07f8
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 d5cb84bc21d5be675127b5a342683c9479064d95
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 47866136666c96480cd4b5857f8f6fe26afe67ef
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 f118b2e39ff43ee2fc1e263a26eb51fdfc5a03c7
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 df298462fe6d86aab712100dd799a2a434f114c8
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 ab72f1e29bba7f7c2fb7eb9c85fe2bba7f89eabc
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 32eab21349844c55bc028dab1f27fb01e346a4e8
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 277c47f61cbdcaacafb48f90ee171694a922c886
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 67959a329436cd827302fb81a4529b365041b86a
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 f1cc63135299cbb946c3131eac125b3c03414848
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 a0695a57c7f084cf8507983e30d5ad0861a6226d
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 3c1d5d4afbf6147d02499be474da5fe711155ec5
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 852ec671223ea2a37a76da9330ab6185f6f35eed
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 9f456f80af4d5f139e842fdefea4df41829c6dee
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 82e53acd7835421acfc6da1f5545589f722e0ecd
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 9c5400c418139beb148bc5afb3d77e789a466202
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 ebb36a050880415fdc2ef480a7558f80ab2cac52
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 49cacb9abf134b5db71fb0a45477bd7fa9db1388
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 0258b415f369c42bda0c876dc49ff5403e15d34c
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 a22289fd97fa587ad7f7973a2757aa57bc086d2d
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 f64f2efb9362ee358bb6f5c85517cedfef2bc014
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 8e49ff3ac98a40ba7df5da0efbe143762cf6fcfc
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 b7c95d51cb79d36aa770afb559ef80448810ad5e
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 9346eef45cdea2dbfc9487c6329f5c6061278365
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 a53be32fefabb3e8825d58a99f59b99d7a4b17f0
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 796a0777b64918b6af98984c41d339d1c4e44485
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 2424056af6d14070bf5be06365984df3be2908af
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 c4b86f73a27690307b8bfbabb756d251a5b73fe7
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 70d008796f45545430c04777fc2560902ea2a119
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 7d3c86427bbc168efc3bb3efd883c52503ca8241
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 8dcefbe4267ea089b5ef073127894a2fd045b7f8
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 36a84ef006cba78e92fd1d3a97363375f3bae845
git diff-tree --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21 a0bc7884cd1bf679fd903122f2315038ccc992c0
git merge-base 73021bc11a37c530a892cbea1bd25d6f986b0c21 f59c47a52e86749d522472a0920dcbb7f7bca88c
git rev-list 73021bc11a37c530a892cbea1bd25d6f986b0c21 ^ec2411e9ca5e3c5984aad397aa0874b00a427f48
git merge-base 73021bc11a37c530a892cbea1bd25d6f986b0c21 ec2411e9ca5e3c5984aad397aa0874b00a427f48
git rev-parse --verify --quiet 73021bc11a37c530a892cbea1bd25d6f986b0c21^{commit}
git rev-parse --verify --quiet ec2411e9ca5e3c5984aad397aa0874b00a427f48^{commit}
git rev-list ec2411e9ca5e3c5984aad397aa0874b00a427f48 ^338a30534863a245347c21108f84868779fbdcee
git merge-base 338a30534863a245347c21108f84868779fbdcee ec2411e9ca5e3c5984aad397aa0874b00a427f48
git version
git remote
git reflog show --format=%gD$'\t'%H$'\t'%gs refs/heads/TEMP/doc/readme-first-5-mins refs/heads/bar refs/heads/bugfix/add-crash refs/heads/bugfix/advance-missing-post-slide-out refs/heads/bugfix/cwd-missing refs/heads/bugfix/gmip-681-discover-fp-logic refs/heads/bugfix/links-405 refs/heads/bugfix/nixos-version refs/heads/bugfix/no-op-rebase-in-fork-point refs/heads/bugfix/show-root refs/heads/bugfix/sync-github-prs refs/heads/chore/annotated-tags refs/heads/chore/check-python refs/heads/chore/ci-snap-publish refs/heads/chore/deploy-on-master refs/heads/chore/discover-main refs/heads/chore/editor-vars refs/heads/chore/focal refs/heads/chore/misc-msgs refs/heads/chore/move-docker-images refs/heads/chore/nixpkgs refs/heads/chore/nixpkgs-dry-run-shallow-fetch refs/heads/chore/remove-travis refs/heads/chore/snap-classic refs/heads/chore/sphinx-docs refs/heads/chore/tox-rectify refs/heads/chore/travis-links refs/heads/circleci-project-setup refs/heads/detect_squash refs/heads/develop refs/heads/doc/readme-first-5-mins refs/heads/docs/asciinema-gif refs/heads/feature/add-from-remote refs/heads/feature/advance refs/heads/feature/anno-github-prs refs/heads/feature/cli-minor refs/heads/feature/delete-on-slide-out refs/heads/feature/go-back refs/heads/feature/hub refs/heads/feature/manage-remote-branches refs/heads/feature/post-slide-out-hook refs/heads/feature/slide-out-invalid-branches refs/heads/feature/squash refs/heads/feature/types refs/heads/fix/ci-minor refs/heads/fix/logging refs/heads/fix/logs refs/heads/fix/nix-no-reviewer refs/heads/fix/rpm-python-pip refs/heads/foo refs/heads/improve/discover-develop-as-root refs/heads/improve/discover-develop-root refs/heads/improve/discover-skip-merged refs/heads/improve/editor-choice refs/heads/improve/error-warn-fmt refs/heads/improve/fork-point-minor refs/heads/improve/help-fmt refs/heads/improve/infer-remote-head refs/heads/improve/prompt-ynq refs/heads/improve/soft-remote-head refs/heads/improve/squash-keep-author refs/heads/improve/traverse-reset-hard refs/heads/master refs/heads/optimize/anno-show-no-verify-branches refs/heads/orphan/brew-package-check refs/heads/refactor/machete-context refs/heads/refactor/slide-out refs/heads/remove/go-branch refs/remotes/origin/bugfix/add-crash refs/remotes/origin/bugfix/gmip-681-discover-fp-logic refs/remotes/origin/bugfix/nixos-version refs/remotes/origin/bugfix/no-op-rebase-in-fork-point refs/remotes/origin/bugfix/sync-github-prs refs/remotes/origin/chore/check-python refs/remotes/origin/chore/misc-msgs refs/remotes/origin/chore/nixpkgs-dry-run-shallow-fetch refs/remotes/origin/chore/remove-travis refs/remotes/origin/chore/travis-links refs/remotes/origin/circleci-project-setup refs/remotes/asford/detect_squash refs/remotes/origin/develop refs/remotes/origin/feature/go-back refs/remotes/origin/feature/hub refs/remotes/origin/feature/manage-remote-branches refs/remotes/origin/feature/types refs/remotes/origin/improve/discover-develop-as-root refs/remotes/origin/improve/infer-remote-head refs/remotes/origin/improve/soft-remote-head refs/remotes/origin/master refs/remotes/origin/refactor/machete-context refs/remotes/origin/remove/go-branch --
git log --max-count=10 --format=%H refs/heads/improve/infer-remote-head
git rev-list 338a30534863a245347c21108f84868779fbdcee ^134695dc7bb6924ca5749020016ac480ce863634
git merge-base 134695dc7bb6924ca5749020016ac480ce863634 338a30534863a245347c21108f84868779fbdcee
git log --max-count=10 --format=%H refs/heads/remove/go-branch
git rev-list 73021bc11a37c530a892cbea1bd25d6f986b0c21 ^2685c77eb3029900bf1eeff4d04f41a754498363
git merge-base 2685c77eb3029900bf1eeff4d04f41a754498363 73021bc11a37c530a892cbea1bd25d6f986b0c21
git log --max-count=10 --format=%H refs/heads/improve/soft-remote-head
git rev-list f59c47a52e86749d522472a0920dcbb7f7bca88c ^53239f284ac29d8e057e726655b35d64fb0490d4
git diff-tree --quiet 53239f284ac29d8e057e726655b35d64fb0490d4 f59c47a52e86749d522472a0920dcbb7f7bca88c
git merge-base 53239f284ac29d8e057e726655b35d64fb0490d4 f59c47a52e86749d522472a0920dcbb7f7bca88c
git rev-list f59c47a52e86749d522472a0920dcbb7f7bca88c ^e6eb556ca7f773a3db3f982013c935bae7401654
git diff-tree --quiet e6eb556ca7f773a3db3f982013c935bae7401654 f59c47a52e86749d522472a0920dcbb7f7bca88c
git diff-tree --quiet e6eb556ca7f773a3db3f982013c935bae7401654 b9b971e1007d3f7c2ddfb309da36c5e8d855ec37
git diff-tree --quiet e6eb556ca7f773a3db3f982013c935bae7401654 a9424c5c1c1a883e7ed776b7c38f6e62d31fdad5
git diff-tree --quiet e6eb556ca7f773a3db3f982013c935bae7401654 0f217af0f99884acba395cee61a83f9c4b8234dd
git diff-tree --quiet e6eb556ca7f773a3db3f982013c935bae7401654 2aae74329af314c95f001dc6c910f231f650563a
git diff-tree --quiet e6eb556ca7f773a3db3f982013c935bae7401654 76543acef3dca7492c28bcd4382f68cdd8e562b2
git diff-tree --quiet e6eb556ca7f773a3db3f982013c935bae7401654 bf802772ca30cc02c85dde555e580a9deb7f6a65
git diff-tree --quiet e6eb556ca7f773a3db3f982013c935bae7401654 3b6c90b1d44333c7678e2052ee96c7cf4a82e23a
git diff-tree --quiet e6eb556ca7f773a3db3f982013c935bae7401654 895073cacd9d7c474381dac5e4486ecec9cc7fda
git diff-tree --quiet e6eb556ca7f773a3db3f982013c935bae7401654 032230e12fbdcc612e417bc69613f51f82a6bacb
git diff-tree --quiet e6eb556ca7f773a3db3f982013c935bae7401654 acc315cb678da3a17a5b8d27a1986770b773037b
git merge-base e6eb556ca7f773a3db3f982013c935bae7401654 f59c47a52e86749d522472a0920dcbb7f7bca88c
git rev-list f59c47a52e86749d522472a0920dcbb7f7bca88c ^79f9edc6825034b45f63c0b7ed8c2c930daa3a74
git diff-tree --quiet 79f9edc6825034b45f63c0b7ed8c2c930daa3a74 f59c47a52e86749d522472a0920dcbb7f7bca88c
git merge-base 79f9edc6825034b45f63c0b7ed8c2c930daa3a74 f59c47a52e86749d522472a0920dcbb7f7bca88c
git symbolic-ref --quiet HEAD
git rev-parse --show-toplevel
ASCII_ONLY=false .git/hooks/machete-status-branch master
ASCII_ONLY=false .git/hooks/machete-status-branch develop
ASCII_ONLY=false .git/hooks/machete-status-branch feature/manage-remote-branches
ASCII_ONLY=false .git/hooks/machete-status-branch feature/go-back
ASCII_ONLY=false .git/hooks/machete-status-branch improve/infer-remote-head
ASCII_ONLY=false .git/hooks/machete-status-branch remove/go-branch
ASCII_ONLY=false .git/hooks/machete-status-branch improve/soft-remote-head
ASCII_ONLY=false .git/hooks/machete-status-branch feature/hub
ASCII_ONLY=false .git/hooks/machete-status-branch refactor/machete-context
ASCII_ONLY=false .git/hooks/machete-status-branch detect_squash
  master  v3.1.1
  │ 
  └─develop  v3.1.2
    │ 
    └─feature/manage-remote-branches  PR #93  v2.16.0
    │ │ 
    │ └─feature/go-back  v2.16.0
    │ │ │ 
    │ │ └─improve/infer-remote-head  # TO BE TESTED  v2.16.0
    │ │   │ 
    │ │   └─remove/go-branch  v2.16.0
    │ │ 
    │ └─improve/soft-remote-head  v2.16.0
    │ 
    └─feature/hub  PR #130  v3.2.0
    │ 
    └─refactor/machete-context  PR #124 (mkondratek)  v3.1.1
    │ 
    └─detect_squash  PR #129 (asford)  v3.2.0

```

Reply:
I don't say that's particularly bad... although the times differ a bit for my sample repository (`git-machete`):

```
real	0m0.636s
user	0m0.368s
sys	0m0.311s
```

vs

```
real	0m0.384s
user	0m0.295s
sys	0m0.116s
```

Reply:
Did you consider using `git cherry` (not to be confused with `git cherry-pick`)? I've never used it myself, but it seems it might be suited for this purpose 🤔

Reply:
`git cherry` isn't really a viable option here, because the actual diffs will be different in a squash-based workflow. For example, it's likely that the entire branch's worth of commits will have been squashed into a single commit on the upstream branch.

`git cherry` is designed to detect cherry-picks or rebases based on diff content, rather than tree state, so it isn't suitable.

While it certainly feels a little gamey to brute-force check the tree status, I think it's an acceptable trade-off for what the feature is providing. ~200-300ms feel reasonable for a status check that's primarily run during interactive workflows and/or with a fair amount of network traffic during push/pull.

If this is a significant performance regression concern I'd be possible to put the feature behind a config flag, but as it's always "disable-able" and I haven't been able to find a case where the diff-tree performance is bad enough to have severe impact, I don't think it's a blocking problem?

Reply:
Musing over this one... I'm generally almost-convinced, lemme just try a small optimization on a separate PR 🕵🏻

Reply:
I benchmarked `cat-file` vrs `diff-tree` because I was interested, `cat-file` is about 10x faster.
I'll fold the `cat-file` implementation into this PR now, no reason not to take the speedup.

https://gist.github.com/asford/4966a816d40d5e0579edbf4e0b9e410f

TLDR:

```
In [27]:
rev_list= !git rev-list HEAD ^HEAD~100
len(rev_list)
Out[27]:
100

In [28]:
%%timeit
matching_tree_commits_via_catfile(rev_list[0], rev_list)
16.2 ms ± 530 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [29]:
%%timeit
matching_tree_commits_via_difftree(rev_list[0], rev_list)
858 ms ± 38.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Great! FWIW, I've quickly rewritten `contains_equivalent_tree` so that a single call to `git log` is done instead of multiple calls to `git diff-tree`:

```diff
diff --git a/git_machete/cmd.py b/git_machete/cmd.py
index ae009e2..433e0b4 100644
--- a/git_machete/cmd.py
+++ b/git_machete/cmd.py
@@ -1139,44 +1139,29 @@ def contains_equivalent_tree(
         f"earlier_revision={earlier_revision} later_revision={later_revision}",
     )
 
-    # git rev-list later_sha ^earlier_sha
-    # shows all commits reachable via later_sha but not by earlier_sha
+    # `git log later_sha ^earlier_sha`
+    # shows all commits reachable from later_sha but not from earlier_sha
     #
     # https://git-scm.com/docs/git-rev-list#_description
-    intermediate_shas = non_empty_lines(
+    intermediate_tree_shas = non_empty_lines(
         popen_git(
             cli_ctxt,
-            "rev-list",
+            "log",
+            "--format=%T",  # full commit's tree hash
             later_sha,
             "^" + earlier_sha,
         )
     )
 
-    # Check later_sha and all the itermediate ancestors of later_sha to
-    # determine if they've an equivalent tree to earlier_sha
-    for intermediate_sha in intermediate_shas:
-        # git diff-tree --quiet
-        # Exits with 1 if there were differences and 0 means no differences.
-        #
-        # https://git-scm.com/docs/git-diff-tree#Documentation/git-diff-tree.txt---quiet
-        diff_result = run_cmd(
-            cli_ctxt,
-            "git",
-            "diff-tree",
-            "--quiet",
-            earlier_sha,
-            intermediate_sha
-        )
-
-        if diff_result == 0:
-            debug(
-                cli_ctxt,
-                "contains_equivalent_tree found",
-                f"earlier_sha={earlier_sha} intermediate_sha={intermediate_sha}"
-            )
-            return True
+    earlier_tree_sha = popen_git(
+        cli_ctxt,
+        "log",
+        "--format=%T",
+        "-1",
+        earlier_sha
+    ).strip()
 
-    return False
+    return earlier_tree_sha in intermediate_tree_shas
 
 
 def create_branch(cli_ctxt: CommandLineContext, b: str, out_of_revision: str) -> None:
```

Reply:
Fantastic, pushing this patch now.

Reply:
Good catch on `--format=%T`! We've identified, what, three ways of doing this same thing now? 
This one is clearly the winner.

Reply:
Haha yes looks like git is an ultimate TIMTOWTDI ;)